langchain/src/tools/outlook/authFlowREST.ts

```diff
@@ -0,0 +1,205 @@
+      `${loginEndpoint}?client_id=${clientId}`,
+      `&response_type=${response_type}`,
+      `&response_mode=${response_mode}`,
+      `&redirect_uri=${redirectUri}`,
+      `&scope=${scope}`,
+      `&state=${state}`,
+    ].join("");
+
+    console.log("Please open the following url to login:");
+    console.log(url);
+    return url;
+  }
+
+  // Function to start the server
+  private startServer(): Promise<http.Server> {
+    return new Promise((resolve, reject) => {
+      const server = http.createServer();
+
+      server.listen(this.port, () => {
+        console.log(`Server listening at http://localhost:${this.port}`);
+        resolve(server);
+      });
+
+      server.on("error", (err) => {
+        reject(err);
+      });
+    });
+  }
+
+  // Function to listen for the authorization code
+  private async listenForCode(server: http.Server): Promise<string> {
+    return new Promise((resolve, reject) => {
+      // Set timeout in case the user fails to login
+      const timeout = setTimeout(() => {
+        server.close();
+        reject(new Error("Timeout"));
+      }, 180000);
+
+      server.on("request", (req, res) => {
+        try {
+          const reqUrl = url.parse(req.url || "", true);
+
+          if (reqUrl.pathname === this.pathname) {
+            const authCode = reqUrl.query.code as string;
+
+            res.writeHead(200, { "Content-Type": "text/html" });
+            res.end("Authorization code received. You can close this window.");
+
+            clearTimeout(timeout);
+            server.close();
+            console.log("Server closed");
+            resolve(authCode); // Resolve the Promise with the authorization code
+          } else {
+            res.writeHead(404);
+            res.end("404 Not Found");
+          }
+        } catch (err) {
+          clearTimeout(timeout);
+          res.writeHead(500);
+          res.end("Server error");
+          reject(err);
+        }
+      });
+    });
+  }
+
+  // get authentication code from user login
+  private async getCode(): Promise<string> {
+    // check credentials
+    if (!this.clientId || !this.redirectUri) {
+      throw new Error("Missing clientId or redirectUri.");
+    }
+
+    const server = await this.startServer();
+    this.openAuthUrl();
+    const code = await this.listenForCode(server);
+    return code;
+  }
+
+  // Function to get the token using the code and client credentials
+  public async getAccessToken(): Promise<string> {
+    // fetch auth code from user login
+    const code = await this.getCode();
+    // fetch access token using auth code
+    const req_body =
+      `client_id=${encodeURIComponent(this.clientId)}&` +
+      `client_secret=${encodeURIComponent(this.clientSecret)}&` +
+      `scope=${encodeURIComponent("https://graph.microsoft.com/.default")}&` +
+      `redirect_uri=${encodeURIComponent(this.redirectUri)}&` +
+      `grant_type=authorization_code&` +
+      `code=${encodeURIComponent(code)}`;
```

Comment:
Could be cleaner to use something like [`URLSearchParams`](https://developer.mozilla.org/en-US/docs/Web/API/URLSearchParams)